### DISCUSSION TITLE
Invalid compiler-message

### DISCUSSION THREAD

**Discussion by AndreasHoffmann2** [dotnet/csharplang#5977] [Language Ideas]
I ran into a compile-error today: "Cannot convert Type 'T' to 'MyDispatchProxy'". You can finde the whole code that created this error below.

The line with the error read:
```C#
var proxyImpl = (MyDispatchProxy)proxyInterface;
```

Seeing the error, I started checking, why this cast would fail at runtime: When the compiler tells me that something is impossible, I do believe it.
After a lot of checking around, I was pretty sure that the conversion is 100% possible. Thus I removed that line and checked at runtime. And I was right: The conversion is perfectly fine at runtime. In my use-case, it will never fail.

Then, I tried some alternatives for the conversion. And all worked perfectly fine:
```C#
var proxyImpl2 = (MyDispatchProxy)(object)proxyInterface;
var proxyImpl3 = proxyInterface as MyDispatchProxy;
if (proxyImpl3 is null && proxyInterface is not null) 
{
	throw new Exception("Unreachable code");
}
if (proxyInterface is MyDispatchProxy proxyImpl4) 
{
	//Use it
} 
else 
{
	throw new Exception("Unreachable code");
}
```
Why does the compiler claim that the direct conversion is impossible?

At least, the error-message should be changed. Something like:
"Converting Type 'T' to 'MyDispatchProxy' like that is currently not supported. Please use another way of converting the reference."

In my opinion, this error should be removed from the language itself.
The language offers 3 ways to convert reference-types:

- cast
- "as"
- "is"

The latter two should be used if the developer expects the conversion to fail and has some alternative code if the conversion fails. A cast on the other hand is the correct choice if a conversion is not expected to fail.

The compile-error breaks this pattern. It removes one perfectly valid option. That's an inconsistency in the language.
Either all three options should be possible. Or all three options should be forbidden.

I prefer allowing all three options of course. This leads to more readable code than the code I have to write now.

Because if my assumption is wrong and this conversion does actually fail at runtime, I want to get an Exception from the exact line where I was wrong. And not a NullReferenceException in some other part of the application that happens to be the first one to use the variable.

The whole example:
```C#
public class MyDispatchProxy : DispatchProxy 
{
	public object Target;
	protected override object Invoke(MethodInfo targetMethod, object[] args) 
	{
		//Do something
		return targetMethod.Invoke(Target, args);
	}
}
public class ProxyFactory 
{
	public static T Create<T>(T target) where T : class 
	{
		var proxyInterface = DispatchProxy.Create<T, MyDispatchProxy>();
		//The next line creates a compile-error: 
		var proxyImpl = (MyDispatchProxy)proxyInterface;
		proxyImpl.Target = target;
		return proxyInterface;
	}
}
```

**Comment by pinkfloydx33** [comment]
The generic `T` (return type of `Create`) does not statically implement `MyDispatchProxy`, there's no generic constraints on your method. The `object` cast would be the correct thing to do.

**Reply by pinkfloydx33** [reply]
You've also not shown how you're calling that method. I assume(?) you are passing `MyDispatchProxy` as the `T`? You may wanna check, because the proxy subtype should usually be the second type argument to `Create` and the first typically an interface that's it's implementing.

**Reply by AndreasHoffmann2** [reply]
Sorry, there was a typo in the call. It's `var proxyInterface = DispatchProxy.Create<T, MyDispatchProxy>();` of course (`MyDispatchProxy` instead of `DispatchProxy`). `T` is the interface, I want to get a proxy for.

**Reply by AndreasHoffmann2** [reply]
About your first reply: 
Why does the compiler force me to cast to `object` first?
`T` is always assignable to `object`. Casting `T` to `object` first is 100% redundant. And since the compiler knows this anyway, there is no reason to force me to write this

Or is there any case, where casting to `object` can actually fail?

**Reply by pinkfloydx33** [reply]
That's how generic casts have always worked.  `T` cannot be statically proven to be your Proxy type. There's no relation between the two at compile time. The cast to `object` is basically you as the developer telling the compiler you know what you're doing.The double cast will be elided by the jjt in most cases (I believe )

**Reply by AndreasHoffmann2** [reply]
> `T` cannot be statically proven to be your Proxy type.

Correct. That's why a cast is necessary in the first place. If it was statically provable, there would be no need for the cast at all (explicit interface-implementations being the exception).

> There's no relation between the two at compile time.

That's the normal case when casting. Take the following code for example:
```C#
interface SomeInterface { }
class SomeClass { }
SomeClass Foo(SomeInterface value)
{
	return (SomeClass)value;
}
```
There is absolutely no relation between `SomeInterface` and `SomeClass` at compile time. But the cast might be successful at runtime anyway. That's why the compiler allows it.
But when I write the exact same thing using generics, the cast is forbidden suddenly:
```C#
SomeClass Foo2<T>(T value) where T : SomeInterface
{
	return (SomeClass)value; //compile-error
}
```
Both implementations are to my knowledge 100% equivalent: If you can use one variable to call Foo, you can use the same variable to call Foo2 and vice versa.

Please give me one good reason, why the cast in the second one should be forbidden. Without quoting the spec please.

Technically speaking, it is definitely correct according to the specs. And the JIT will probably catch it and there will be no performance-drawbacks. But that's not the point.
It's about specifying a language that behaves consistently and transparently for the users (i.e. the developers).

**Reply by VolodymyrBS** [reply]
@AndreasHoffmann2 
I agree that compiler behaviour looks too restrictive here. but I think that such behaviour was chosen because there are more chances for error in a generic version. T could be any type that implements `SomeInterface`.

so in the first example interface is treated as an `object` type that could be cast to anything.
In the second example, T is treated concrete type/ and there are many cases where cast would fail. 

And still, language leaves a hole to do the unsafe conversion if you are 100% sure it is safe

**Reply by AndreasHoffmann2** [reply]
The chances of the cast failing are exactly the same in the both versions. 100 % equal, since both methods accept the exact same parameters.

Casting to object first is more like a hack to make the stupid compiler shut up.

**Reply by jnm2** [reply]
@AndreasHoffmann2 If you were casting to an interface, that would be one thing, but you're casting to a class. If you have three classes, `Animal`, `Cat : Animal`, and `Dog : Animal`, you can't cast a `Dog` instance to `Cat` directly without casting to the base type `Animal` first. `T` is a type that "derives" from `object` in a sense, so it is not directly castable to `MyDispatchProxy` without an intermediate `object` cast just like how `Dog` can't be cast to `Cat` without an intermediate cast.

**Reply by AndreasHoffmann2** [reply]
Why "can't" I cast directly?
Look at what is happening during a cast:
You have a statically typed reference of Type A. And you have an object of Type B, where the reference points to. The language guarantees that B is a subtype of A (or equal to it). 
During a cast to type C, you tell the runtime to check if B happens to be a subtype of C as well.
If this is the case, you get a new statically typed reference of Type C. If not, you get a CastException.
Thd object itself remains totally unchanged during the process.

Where is the point of converting the refence-type to Object first? Even after the conversion of the reference-type, the object ist still of type B.
And the check if B is a subtype of C is exactly the same.

**Reply by jnm2** [reply]
@AndreasHoffmann2 The reason why you can't is because casting `T` to `MyDispatchProxy` is currently like casting `Cat` to `Dog`. You can't do either one directly. To propose a change to this, either you need to make the case that the compiler should allow casting from `Cat` to `Dog` without an intermediate `Animal` (or `object`) cast, or you need to make the case that `T` is not "like" a subtype of `object` in this way.

**Reply by jnm2** [reply]
Put another way, this is the current inheritance situation. In both cases, you have to cast up to the base type before casting down to the other sibling subtype.
```mermaid
classDiagram
    object <|-- T
    object <|-- MyDispatchProxy
    Animal <|-- Cat
    Animal <|-- Dog
```

**Reply by AndreasHoffmann2** [reply]
You have to distinguish between the reference type (determined at compile-time) and the object-type (determined at runtime). These are two entirely different things. Really!

The following image shows, what is going on when casting:
- `new Dog()` creates an new instance of `Dog` in the memory and returns a reference to it of type `Dog`. 
- You can call `d.Bark()` if you want to.
- Since every dog is an `Animal`, you can also assign a reference of type `Animal` to the object of type `Dog`. 
- The object in memory is still a `Dog`. The cast of the reference to `Animal` does not alter the object at all. Also the method `Bark()` is still there. You just cannot access it using a reference of type `Animal`.
- And the cast to `Cat` will always fail. Because the object in memory is actually a `Dog`, not a `Cat`.
- That's the reason, why the compiler does not let you convert a reference of type `Dog` to a reference of type `Cat`: it knows that this is impossible.

![image](https://user-images.githubusercontent.com/8135241/161799514-d8e5bf01-1f22-4f67-8c5a-aabd4e567bb7.png)

Apply this to the `DispatchProxy`.
The method `Create<T, MyDispatchProxy>()` generates an object with two properties:
- it is a subclass of `MyDispatchProxy`
- and it implements the interface T.

After creating the object with both properties, the method returns a reference of type `T`. But the object itself is a subtype of both: `T` and `MyDispatchProxy`.
And all I do is ask the runtime to give me a reference of type 'MyDispatchProxy` to that object. But the compiler does not let me do that.

PS: Please be aware, that castings on reference types as described are entirely different from the conversion of value-types:
```C#
float f = 1.1;
int i = (int)f;
```
Here, the value is actually altered. You cannot have an `int` that is actually a `float` or vice versa. Unfortunately, the designers of the language chose the exact same syntax for two entirely different things.

**Reply by jnm2** [reply]
@AndreasHoffmann2 I'm fully aware that these are two different things. However, the static typing situation does not do flow analysis and could never do perfect flow analysis to track the runtime types. Some simplistic static typing rules must apply, and I think you still have to choose (and support) one of the two static typing modifications I thought of above.

**Reply by jnm2** [reply]
Or, add your vote for intersection types where `MyDispatchProzy & T` would be an expressible return type: https://github.com/dotnet/csharplang/discussions/399

**Reply by AndreasHoffmann2** [reply]
I'm not asking anybody to do flow analysis. That's what I'm there for.
Can you explain, why the following code does compile and run without any issue?
```C#
IDispose d = DispatchProxy.Create<IDispose, MyDispatchProxy>();
MyDispatchProxy m = (MyDispatchProxy)d;
```
That's the exact same situation where you are claiming that I am casting a `Cat` to a `Dog`.

I chose `IDispose` only because it is the first interface that came to my mind.

**Reply by jnm2** [reply]
It initially sounds reasonable to me to add a special case so that `T` can be directly explicitly cast to any type so long as `T` doesn't have a type constraint on it. With your use case as evidence. We should also try to think of ways in which new pits of failure could appear, and possible mitigations if there are any. Reasons not to make a change would include:
- The workaround is simple and light once you discover it (and at least some people discover it immediately through gut expectation that an intermediate `object` cast often solves this kind of thing)
- This doesn't sound like something that would affect the average developer even once in an average year.

If I remember right, there have been other places where initially you had to cast `T` to `object` before using it, e.g. in pattern matching, which was later relaxed. That prior art probably helps.

**Reply by jnm2** [reply]
> That's the exact same situation where you are claiming that I am casting a Cat to a Dog.

Please notice that I didn't make that claim. I described types from the perspective of today's static typing rules which follow a specific structure of having `T` be a subtype of `object` and don't yet have a special deviation added to aid your use case. That explains why today's language is consistent with the priorities that have been considered up till now, which is also the necessary starting point for considering making a change.

**Reply by CyrusNajmabadi** [reply]
The simple answer here is that we don't think it's good to allow cross-casts across an inheritance hierarchy.  It often signifies a potential unintentional bug in the user code.  As such, we believe it's safer and more maintainable for the code to be explicit in terms of how it casts along the hierarchy.  So that means casting 'up' first to a common base type, then casting 'down' to the final destination type in the different part of hte hierarchy.

@AndreasHoffmann2 we could do what you're asking.  But we choose not to because we think there is value in having the code be explicit in intent here and not accidentally doing something that might more often than not be *unintended*.

**Reply by AndreasHoffmann2** [reply]
> Or, add your vote for intersection types where `MyDispatchProzy & T` would be an expressible return type: #399

Sounds interesting. I'll have a read sometime.

> We should also try to think of ways in which new pits of failure could appear

I'll have a think about it. Currently I don't see any new pitfalls. 
But I see one more advantage: the workaround via `object` stops the compiler from determining if the cast is possible in some cases:
Just assume `T` is constrained with `where T : Dog`. The routined developer will automatically write the cast to `object` first to reach the target type. This way, a cast to `Cat` will be allowed. Even though it is guaranteed to fail at runtime. Without the intermediate step via `object`, the compiler would be able to catch the issue and display an error.

> The workaround is simple and light once you discover it

I see the issue more in the loss of trust in the compiler. For me, the compiler is always right. And it has always been in the past. And I encountered the message "X cannot be converted to Y" a lot of times actually.
If the compiler tells me this, I believe it and search my error. Questioning the compiler does not feel right.

> This doesn't sound like something that would affect the average developer even once in an average year.

Definitely not.

**Reply by CyrusNajmabadi** [reply]
> automatically write the cast to object first to reach the target type.

But here's the thing.  Then taht cast is in source.  It's easy to examine and go: ok, they user is casting up/down the hierarchy.  Do these two steps make sense?  e.g. the same thing you'd be asking if you saw  `(Dog)(Animal)new Cat()`.

> The routined developer will 

Here's the thing though, we've had these rules and behaviors for almost 20 years now.  We're not seeing any routine problem here.  I'd be more amenable to these arguments back in 2002.  At this point, the rules have become tried/true and i def take a more conservative stance of "if it ain't broke".

> And I encountered the message "X cannot be converted to Y" a lot of times actually.
If the compiler tells me this, I believe it and search my error. Questioning the compiler does not feel right.

I don't get it.  A 'cast' is exactly about telling the compiler waht you think is write.  But it also comes with conditions to satisfy.  We don't *always* believe you.  And we think even when you're telling us this, there are good cases where you are likely still wrong and making a mistake.  So we have the rules setup to form a balance of how explicit we think you need to be and how much danger we want to help you avoid.

**Reply by AndreasHoffmann2** [reply]
> As such, we believe it's safer and more maintainable for the code to be explicit in terms of how it casts along the hierarchy.

Can you give me another example, where I am forced to specify the way up the object-hierarchy in order to cast a reference-type? I don't see any.
I can write all of the following:
```C#
interface ISomeInterface { }
class SomeClass { }
void Foo(ISomeInterface i)
{
	bool isSomeClass = i is SomeClass s1;
	SomeClass s2 = i as SomeClass;
	SomeClass s3 = (SomeClass) i;
}
void Foo2<T>(T i) where T : ISomeInterface
{
	bool isSomeClass = i is SomeClass s1;
	SomeClass s2 = i as SomeClass;
	SomeClass s3 = (SomeClass) i;
}
```
All lines are doing kind of the same and will cast `i` to `SomeClass`. Only the last one is rejected by the compiler.

**Reply by jnm2** [reply]
@AndreasHoffmann2 `T` where `T : ISomeInterface` is crucially different from `ISomeInterface`, since interface types have casting rules unique to them which generic types don't today. The question we're considering is whether generic types should be castable in all the same ways as interface types, or at least in this one specific way. It's not obvious to me that the answer is unconditionally "yes."

**Reply by jnm2** [reply]
> Can you give me another example, where I am forced to specify the way up the object-hierarchy in order to cast a reference-type? I don't see any.

@AndreasHoffmann2 Yes, here's another example. The compiler will force you to cast `Cat` to `Animal` (or `object`) before casting to `Dog`, even though it's possible that *at runtime* Cat inherits from Dog or Dog inherits from Cat. (Where the implementation assemblies at runtime are newer than the reference assemblies that were used to compile against.)

Your argument has been: "it's not impossible for this cast to succeed at runtime, so the compiler shouldn't make me be more explicit." But whether the cast can possibly succeed at runtime hasn't been the total criterion. If it was, then you could cast `Cat` to `Dog` and vice versa: it's not impossible that a runtime type can be both Cat and Dog. (Though likely in a real-world case, with names that aren't conceptually mutually exclusive.)

**Reply by CyrusNajmabadi** [reply]
> All lines are doing kind of the same and will cast i to SomeClass. Only the last one is rejected by the compiler.

The last cast is not a cast of `ISomeInterface`, it's a cast of T, and we take the view that T is much broader in what it could be at runtime, including branches of the inheritance hierarchy unrelated to SomeClass.  

To make them equivalent in our minds, you'd need to do:

```c#
SomeClass s3 = (SomeClass)(ISomeInterface)i;
```

Again, this is simply a POV.  You may think this adds no value.  When we did the design of generics we did not agree, and we felt it was appropriate to call out the higher level of perceived danger/mistakes that could happen here, and we felt it was not only suitable, but *appropriate* and *desirable* for the code to be extra explicit on how this type fit into the inheritance hierarchy.

**Reply by AndreasHoffmann2** [reply]
> e.g. the same thing you'd be asking if you saw `(Dog)(Animal)new Cat()`.

That's what happens if you force the developer to routinely cast up the hierarchy first. The compiler can easily tell the developer that `(Dog)new Cat()` will always fail.
With the intermediate step of casting to `Animal`, you take that information away. And you let the developer run code that is guaranteed to fail 100% of the time.

> We don't always believe you.

When it comes to casting of reference-types, the compiler does actually believe almost everything as shown in my last post. Only one of the six almost identical statements is rejected. That looks an awful lot like a bug in the spec and not like well thought out feature.

**Reply by jnm2** [reply]
> Only one of the six almost identical statements is rejected. That looks an awful lot like a bug in the spec and not like well thought out feature.

@AndreasHoffmann2 There are some assumptions here that are not self-evident and which I don't share as expectations. Plenty of times I expect there to be possible combinations of C# language features which lead to the natural emergence of a corner case which stands out, without implying anything about the design quality.

Thinking pragmatically about the design *does* take into account the amount of bliss or stress people might feel in common paths, but it's probably not the most important factor.

**Reply by AndreasHoffmann2** [reply]
> The question we're considering is whether generic types should be castable in all the same ways as interface types, or at least in this one specific way.

When it comes to casting, the concept is: "Let the developer do it at his own risk. Except when the compiler can prove the developer wrong at compile time.".

This rule should be applied to generic types as well. And since the compiler knows less about the actual type of T, more things become possible. But only if `T` is a reference type. There is a good reason in the spec, why casting of generic types, that might be a value- or a reference-type is a lot more strict.

> Where the implementation assemblies at runtime are newer than the reference assemblies that were used to compile against.

I think, we should not go down that road. With this in mind, the compiler should not even reject accessing an unknown method. Because it might exist in some newer version, the code will be executed with.

**Reply by AndreasHoffmann2** [reply]
> To make them equivalent in our minds, you'd need to do:

There is absolutely no benefit in the intermediate step. The object `i` will either be `SomeClass` or not. No matter if you put obvious information in between or not.

**Reply by jnm2** [reply]
> I think, we should not go down that road. With this in mind, the compiler should not even reject accessing an unknown method. Because it might exist in some newer version, the code will be executed with.

Exactly. And this is not in the least an uncommon situation; the BCL and NuGet assemblies are newer at runtime than compile time all the time. This demonstrates that there is a line somewhere. It can't be simply, as you said:

> "Let the developer do it at his own risk. Except when the compiler can prove the developer wrong at compile time."

Treating `T` as a subtype of `object` is another such line. The lines have to be considered on their own merit and can't be dismissed by a rule of thumb like "let the developer do it if the compiler can't prove that it will fail." That's a nice motivating ideal, but now that we come to consider the pragmatics of the situation, we shouldn't be trying to shortcut the process by jumping to one of the motivating ideals.

**Reply by AndreasHoffmann2** [reply]
The most depressing part for me: After this conversation, I might invest the effort to create a proposal only to have it closed within minutes with some reference to an older discussion about the same thing.
But the fact that nobody pointed to that discussion (yet) makes me a tiny bit optimistic.

**Reply by AndreasHoffmann2** [reply]
> Treating `T` as a subtype of `object` is another such line.

Currently, I am more in favour of treating T as a subtype of whatever is the where-clause is:

- no where: No change to today's behaviour.
- `object` or below: Apply the rules as if `T` was exactly this type.
- `interface`: as interfaces today (casting to all types is allowed)
- `class`: same as interface.

But I did not think the details through yet.

**Reply by jnm2** [reply]
No `where` and `where T : object?` are currently semantically identical, I believe.

**Reply by AndreasHoffmann2** [reply]
> "let the developer do it if the compiler can't prove that it will fail." That's a nice motivating ideal

That's not only a nice motivating ideal, but exactly how the language behaves today. Except this one case that made me start this discussion.

**Reply by jnm2** [reply]
It's not exactly how the language behaves today, or else (as I observed) it would allow direct casts between Cat and Dog and (as you observed) it would allow calling as-yet-undefined methods because the compiler can't prove that no such method will exist at runtime. There are at least two exceptions to that ideal, one of which is obviously good to you (treating runtime metadata as identical to compile-time metadata) and one of which is obviously bad to you (treating `T` as a subtype of `object` without special affordances).

If I'm not sure that the latter exception is obviously bad, we'll have to talk about it in its own terms. Restating the ideal doesn't resolve the difference in how we see one of the existing exceptions to it.

**Reply by AndreasHoffmann2** [reply]
> No `where` and `where T : object?` are currently semantically identical, I believe.

`where T: object?` is a compile error, because constraining to `object` is not allowed.

**Reply by AndreasHoffmann2** [reply]
I am surprised that you insist that outdated libraries are a valid use case. C# is a strongly typed language and the compiler has to assume that it's current knowledge of what is possible and what not will apply to the runtime as well. If we gave up that assumption, we could just rename the language to javascript :-)

But even for this case, C# has something to offer: declare all variables as `dynamic` and hope that the libraries will be adjusted to the code in the future.

**Reply by jnm2** [reply]
> I am surprised that you insist that outdated libraries are a valid case.

@AndreasHoffmann2 I don't recognize something I said in this sentence. If it helps clarify what I said, it's fairly common to target `net5.0` and run against .NET 6 runtime libraries, for instance. The same was always true in .NET Framework 4.x days. With NuGet, it's constantly the case that a library that references a library in another package will actually run against a newer version at runtime because the app had a separate dependency path which upgraded it.

**Reply by jnm2** [reply]
> But even for this case, C# has something to offer: declare all variables as dynamic and hope that the libraries will be adjusted to the code in the future.

@AndreasHoffmann2 It likewise has something to offer in your case: adding `(object)` into your code.

**Reply by jnm2** [reply]
I think we should focus on ergonomics and considering pitfalls, like what @CyrusNajmabadi was discussing. The motivating rule of thumb is fine but it's not the only factor and we keep going into the weeds talking about it.

**Reply by CyrusNajmabadi** [reply]
> That's what happens if you force the developer to routinely cast up the hierarchy first.

Again, we haven't seen this be an issue in 20 years.  As such, our view woudl likely be that this is a sensible behavior for the language that does more help than harm.  

As i have stated *explicitly*, there is nothing conceptually wrong with your perspective here.  It just doesn't align with our values on where we think the balance should be between guardrails vs permissiveness.  Absent strong community feedback that the current status quo is an issue, we would not likely cahnge this.

**Reply by CyrusNajmabadi** [reply]
> That looks an awful lot like a bug in the spec 

It is not.  I was there when it was designed and this was very intentional.  In practice, i've personally found this valuable and it has prevent unintented bugs on my part.  So i'm personally of the opinion that this is valuable and should stay absent large amounts of community feedback to the contrary.

**Reply by CyrusNajmabadi** [reply]
> When it comes to casting, the concept is: "Let the developer do it at his own risk. Except when the compiler can prove the developer wrong at compile time.".

That's not true.  That may be *your* belief, but that is not eh design of our language.  Again, ultimately the language decides where it wants to land on the spectrum of these sorts of rules.  You believe that we should always let the user do it at their own risk.  We do not.  And that's been a core part of hte design of the language for 20 years.  We are unlikely to change our philosophy there after 1 user report in 20 years as we feel that this balance has been appropriate and has done more good than harm.

Again, if this were 20 years ago, i might feel there was more reason to debate this.  Similarly, if this was a constant piece of feedbakc we heard, or this garnered a large amount of community feedback that our balance here is a problem, then we would more amenable to relaxing things.

But we are unlikely to change things based on a single report from a community member (out of millions) that our balance doesn't align with their personal views on how the language should work.

**Reply by CyrusNajmabadi** [reply]
> There is absolutely no benefit in the intermediate step. 

There is no benefit to *you*.  There is benefit to *me*.  :)

> The object `i` will either be `SomeClass` or not. No matter if you put obvious information in between or not.

It's non-obvious to me.  And having that information helps me understand that the potential cross hierarchy cast is intentional and the code is stating it thinks it is safe because of the path specified explicitly through the casts.

**Reply by CyrusNajmabadi** [reply]
>That's not only a nice motivating ideal, but exactly how the language behaves today. Except this one case that made me start this discussion.

This is not true.  It is not how the language behaves today.  We block things we know will always fail.  We allow things that may fail, sometimes with a single cast, sometimes with a dual cast.  This was intentional and we believe there is value here.  I totally get that you do not believe in that.  But ultimately, we look at our history, our community and our own experiences to make a call on multiple approaches that can work.  In this case, all three of those have led us to be very comfortable and happy with the decision here, and the last 20 years of experience and feedback have not caused us to revise that.

**Reply by CyrusNajmabadi** [reply]
> The most depressing part for me: After this conversation, I might invest the effort to create a proposal

Please do not create a proposal.  It goes against the rules and policies for our repo.  As per our rules here: https://github.com/dotnet/csharplang#c-language-design

> For new feature proposals, however, please raise them for [discussion](https://github.com/dotnet/csharplang/labels/Discussion), and only submit a proposal as a pull request if invited to do so by a member of the Language Design Team (a "champion").

Proposals happen when we and the community decide this is something we would want to change.  To get there we need tehse discussions to flesh out the space and actually see if there is broad community support here.

Thanks!

**Reply by AndreasHoffmann2** [reply]
> You believe that we should always let the user do it at their own risk. We do not.

Can you please show me one other example where the casting of a reference type is forbidden even though it might be successful at runtime?
(except running the code against a different library that defines the same types in a different hierarchy).

Casting up the hierarchy before casting down is a very risky operation, since has the potential to convert a compile-error into a guaranteed runtime-error.

> I was there when it was designed and this was very intentional.

Can you elaborate, why the cast across the hierarchy is allowed with the "as"-statement then?
That's even more risky. Because the cast does not fail on the spot but is likely to generate a NullReferenceException somewhere else.

**Reply by AndreasHoffmann2** [reply]
> For new feature proposals, however,

Sorry, I did not get to this point yet and would definitely have stopped there of course.

**Reply by AndreasHoffmann2** [reply]
One more thought:
```C#
void Foo<T>(T dog) where T : Dog {
	Cat c1 = (Cat)(object)dog;
}
```
This does compile just fine. Even though it is guaranteed to fail at runtime. 
If the compiler treated `T` as the type `Dog` while evaluating the casts, the developer would write:
```C#
void Foo<T>(T dog) where T : Dog {
	Cat c1 = (Cat)dog;
}
```
And the compiler could easily detect that this is in fact impossible.

**Comment by ufcpp** [comment]
If you are sure that `T` inherits from `MyDispatchProxy` but the compiler could not, you can use `Unsafe.As`:

```cs
		var proxyImpl = Unsafe.As<T, MyDispatchProxy>(ref proxyInterface);
```

**Reply by AndreasHoffmann2** [reply]
I don't think that an unsafe conversion is the correct way to go to cast from an interface to the implementation.

**Reply by jnm2** [reply]
@ufcpp Does this have an advantage over `(MyDispatchProxy)(object)proxyInterface`?

**Reply by ufcpp** [reply]
Yes, it's faster because of omitting runtime type check.

**Comment by VolodymyrBS** [comment]
this is by design

Take a look at cast specification
  -  [Cast expression](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#1187-cast-expressions)
  - [Explicit reference conversions](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/conversions.md#1035-explicit-reference-conversions)

At the second link you can find that it is possible to cast object to any class or cast Type S to any T if T is subclass of S.

In you example you are trying to cast generic T to MyDispatchProxy, but you can't do that becouse of second rule. But after casting to object first rule came into play

**Reply by AndreasHoffmann2** [reply]
That's correct. The compile-error is in line with the spec. It is the specification, that should be changed: a cast should only create a compile-error if the cast ist guaranteed to fail at runtime.

That's the developer's perspective. He want to tell the compiler something about a reference that it does not know. And risk a CastException if he is wrong.

If the compiler can prove the developer wrong at compile-time, it should do so. That's its job.

But it should never evee reject a cast that has the possibility of going through at runtime. That is against the purpose of casting.

All the cases mentioned in the spec follow this rule in non-generic code: the compiler does only reject casts that are truly impossible.

But when dealing with generics, these rule do suddenly start to reject casts that can and will work at runtime. I provided an example for that above.
And that is no desirable behavior.

**Reply by VolodymyrBS** [reply]
I got your point.

But I still think that compiler is right here. Compiler not only stops you from writing 100% wrong code, but also from writing dungerous code.
Let's take away `DispatchProxy.Create` becouse it's has lot of "runtime magic" and take this simpler code
```csharp
class MyClass
{
} 

void Do<T>() 
    where T : new() 
{
    var mydata = (MyClass) new T() ;
} 
```

should this code compile? I think no, becouse there ais big chanse that cast will fail. T could be any type with default but set of such types is bigger that set of classes that derive MyClass, so there are more cases when cast will fail. 

Code with `DispatchProxy ` looks the same for compiler. It  see some generic type T that could be any class with no connection to MyDispatchProxy

**Reply by AndreasHoffmann2** [reply]
That's an interesting example that made me question myself. And then I tries the following:
```C#
class SomeClass { }
class SomeSubclass : SomeClass { }
SomeSubclass Foo() 
{
	return (SomeSubclass)new SomeClass();
}
```
This casting is guaranteed to fail. And yet the compiler accepts it. Because the concept of a cast is "the developer is always right". Unless the compiler is 100 % sure that the developer is actually wrong. In this case, the developer is 100% wrong, but nobody invested the effort yet to add this case to the spec. And when in doubt, a cast should be allowed.

Another argument are the following methods Foo2 and Foo3. The methods are 100% equivalent: If you can call Foo2 with a variable, then you can call Foo3 with it as well. And vice versa. And the results will always be the same.
But Foo2 compiles and Foo3 does not.
```C#
interface ISomeInterface { }
class SomeClass { }
class SomeSubclass : SomeClass { }
SomeSubclass Foo() 
{
	return (SomeSubclass)new SomeClass();
}
SomeClass Foo2(ISomeInterface value) 
{
	return (SomeClass)value;
}
SomeClass Foo3<T>(T value) where T : ISomeInterface 
{
	return (SomeClass)value; //compile error
}
```

**Reply by Richiban** [reply]
To be honest, I think the aberration here is the fact that this works:

```cs
interface SomeInterface { }

class SomeClass { }

SomeClass Foo(SomeInterface value)
{
	return (SomeClass)value;
}
```

The casting rules between interfaces and classes seem to behave differently, but otherwise I would expect to be able to cast *up* or *down* the object hierarchy, not *across*. Put in other words, for any expression `(A)b` where I am casting from a type `B` to a type `A` in order for this to compile `A` must be either a subtype or supertype of `B`. You can't just cast from any type to any other type except, if you really, really know more than the compiler at that point, by simply casting up to `object` first. 

This is exactly how I'd expect it to work; the fact that cross-casting is allowed from an interface to pretty much any class is very unexpected to me.

**Reply by AndreasHoffmann2** [reply]
> You can't just cast from any type to any other type except, if you really, really know more than the compiler at that point

The whole point of casting down the object hierarchy is exactly the same: Tell the compiler something it does not know at that point. That's what a casting is there for.

Casting to object first does not give the compiler any additional information. So why force the developer to do it?
It's more of a hack since the developer knows that the compiler is too stupid to remember what the reference was before casting.

**Reply by Richiban** [reply]
It's not so much "forcing the developer to do it" as "the compiler can't stop you". Casting any type up to `object` is always legal, and casting `object` down to any other type is also legal. I still think that casting across the object tree should be illegal, and if I was designing my own language I wouldn't allow it with interfaces either.

**Reply by FaustVX** [reply]
@AndreasHoffmann2 
The difference is
```csharp
interface ISomeInterface { }
class SomeClass1 :  ISomeInterface { }
class SomeClass2  : ISomeInterface { }

SomeClass1 Foo2(ISomeInterface value) 
{
	return (SomeClass1)value;
}
SomeClass1 Foo3<T>(T value) where T : ISomeInterface 
{
	return (SomeClass1)value; //compile error
}

Foo3(new SomeClass2());
// It's the same as calling
SomeClass1 Foo3(SomeClass2value) 
{
	return (SomeClass1)value; // compiler error
}
```

**Reply by Joe4evr** [reply]
@AndreasHoffmann2 The thing about your `Foo2` is that, [unless `SomeClass` is sealed](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQCYBGAWAChLYEMsIBjVAAgEkBlAezFVZlvT0mzAN7MAvpQDOqCABtUAE2bFmXHgGE5EKVNETKlAiQBsKosw2URlZnZWm13VFp16AYp05EAFB2d8AkIsyPIArqgAlMy29jYU9okqAOzMPuou2rqRoXIRANyxdpIU4kA==), the compiler has to assume `SomeOtherSubclass : SomeClass, ISomeInterface` *can* exist, in which case the cast would succeed, so the conversion is allowed.

`Foo3` on the other hand, involves a type parameter [which has its own behavior defined in the spec](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/conversions.md#1038-explicit-conversions-involving-type-parameters):
> The above rules do not permit a direct explicit conversion from an unconstrained type parameter to a non-interface type, which might be surprising. The reason for this rule is to prevent confusion and make the semantics of such conversions clear.

**Reply by AndreasHoffmann2** [reply]
Thank you for the reference to the spec. They even added an example why they did not allow that. But the example is for an unconstrained reference type. And it makes total sense there to make the developer show the compiler that he wants to cast a reference type. The issue here is, that conversions (e.g. float to int) and casts of the reference type use the same Syntax in C#, but are entirely different things.
But in my case, it's not an unconstrained reference type. The compiler does very well know that T has to be a reference type. There are no conversions involved here.
The reasoning from the spec does thus not apply to this case.

**Reply by CyrusNajmabadi** [reply]
> It is the specification, that should be changed: a cast should only create a compile-error if the cast ist guaranteed to fail at runtime.

This is a perfectly fine belief to have.  Truly, that could be the principle we took in the language.  But it is not aligned with our beliefs on the guardrails we think the language should have.  More often than not, we think in these generic contexts if we allowed this, one would end up with broken code and then be unhappy that nothing helped prevent it.  As such, we require the more verbose, but higher-clarity/self-document approach that is in *our opinion* better overall.

Fundamentally, this is a philosophy discussion.  Where on the side of protection/permission do we land, and where would you prefer we land.  In this case, the error is intentional and exists for this exact purpose.  The code to address it also matches what we believe is appropriate for clearer and more maintainable code.  So all this is working out as we wanted *for our* set of language priorities.  

Again, i get that you may not like this, and I fully agree that your POV is consistent and reasonable given you set of values toward these language aspects.  It's just that your POV doesn't align with ours, and I do not think it is likely we would change any of that here absent a *huge* upswelling of support from the ecosystem and community stating that these guardrails are problematic and a significant impediment.  We've had this design for nearly 20 years now, and have almost never heard this.  So i'd say we're *much* more likely to stick with the status quo here.
